Optimize the calculation of trade networks

TomWerner opened the pull request:
Rather than recomputing the pathing on the fly every time we need a city's luxury count, we can compute the whole network and cache it, and only invalidate it when roads are built (and eventually pillaged) and when cities are built and destroyed.

This cuts the runtime of the load conquest scenarios test from ~53 seconds to ~16 seconds on my (admittedly slow) laptop.

TomWerner committed:
Optimize the calculation of trade networks

Rather than recomputing the pathing on the fly every time we need a city's luxury count, we can compute the whole network and cache it, and only invalidate it when roads are built (and eventually pillaged) and when cities are built and destroyed.

This cuts the runtime of the load conquest scenarios test from ~53 seconds to ~16 seconds on my laptop.

esbudylin commented on C7Engine/AI/Pathing/TradeNetwork.cs:
```diff
@@ -0,0 +1,80 @@
+
+		public void AddTile(Tile t, Player p) {
+			tiles.Add(t);
+			if (t.Resource != Resource.NONE && t.OwningPlayer() == p && p.KnowsAboutResource(t.Resource)) {
```
When calculating resource counts, you check if a player knows about a resource. Doesn't this mean this data should be invalidated when discovering a new tech? Personally, I think it would be better to keep this check out of TradeNetwork.

TomWerner replied:
Good call, moved the check out.

esbudylin commented on C7Engine/C7GameData/Terraform.cs:
```diff
@@ -6,14 +6,20 @@
+				tile.overlays.Add(TerrainImprovement.road);
+
+				// The trade network needs to be recomputed whenever roads are added.
+				player.InvalidateCachedTradeNetwork();
```
I think a better place for invalidating the trade network cache when building a road might be the Add method of the TileOverlays class. This class will likely also handle cache invalidation for road pillaging when we implement it.

TomWerner replied:
Good call - this also made me realize we need to invalidate all the players trade networks, due to potential transitive effects.

esbudylin reviewed (commented):
Nice work! As a TODO, I guess we'll also need to invalidate cache on war declarations and creation/destruction of airports and harbors?

TomWerner committed:
Address review comments

esbudylin commented:
Can you explain how your code handles border changes related to cultural expansion? For example, if a city's borders expand to include a roaded tile with a resource.

TomWerner committed:
Review comments part 2

TomWerner commented:
> Can you explain how your code handles border changes related to cultural expansion? For example, if a city's borders expand to include a roaded tile with a resource.

Oh good catch, that was a bug. Fixed.

esbudylin commented:
I think there is still an unhandled case here. How does the game behave if civilization A culture-flips a resource tile from civilization B? Will civilization B's network cache be invalidated? I think some of the cache invalidation complexity can be reduced by storing all resources accessible via the trade network in the cache, and checking for tile ownership only when accessing this data. My guess that it shouldn't be too bad performance-wise.

TomWerner committed:
Use a single trade network object for all players

This ensures we always recalculate all the trade networks whenever we invalidate it. Any time one player's trade network would be invalidated we need to do so for all of them - player A's border changing could influence what resources player B has. Player A building a city could impact the transitive road network player B depends on. Ditto with player A building a road, etc etc.

Doing this required pushing some access to the `EngineStorage.gameData` singleton up a few levels because we indirectly look up resources when loading scenarios. This happens when doing the initial citizen assignment, since we need to know how happy the city is to properly place the citizens. In just about every case pushing the call up improved the code, either fixing a latent issue where the singleton was accessed without proper locking, or making it easier to track the lock.

In [this review comment](https://github.com/C7-Game/Prototype/pull/745#issuecomment-2952690460) we discussed ignoring resource owners until it is time to look up resources, but I went with this route because I realized we always need to recalculate all the players' trade routes. So we don't need to ignore resource owners.

I also made the trade network more of a class, which more easily allows for checking whether the player knows about a resource. However this time I did it at the time of checking for resource access, not when building the trade network, so there's no need to invalidate when a tech is discovered (until techs like navigation).

TomWerner commented:
> I think there is still an unhandled case here. How does the game behave if civilization A culture-flips a resource tile from civilization B? Will civilization B's network cache be invalidated? I think some of the cache invalidation complexity can be reduced by storing all resources accessible via the trade network in the cache, and checking for tile ownership only when accessing this data. My guess that it shouldn't be too bad performance-wise.

Yes, that's a good catch. I realized that we always need to recalculate the trade network of all the players after any network is invalidated, so I switched to storing it all in one class. This ensures there's no cases of mismatches, and with a few nicer methods I was able to pull the logic for resource knowledge into the class, but at lookup time, not network building time. So it still fixes the issue of automatically updating things once a tech is learned.

Sorry for the delay on this, I was most of the way through the player setup logic and wanted to finish that up before this.

esbudylin commented on C7Engine/C7GameData/City.cs:
```diff
@@ -137,21 +137,17 @@ public List<CityBuilding> GetBuildings() {
 
-		public IEnumerable<IProducible> ListProductionOptions() {
-			HashSet<Resource> accessibleResources = GetAccessibleResources();
+		public IEnumerable<IProducible> ListProductionOptions(GameData gameData) {
```
Non-blocking: Why do methods of the City class accept gameData as an argument? Can't they just access it directly via EngineStorage?

TomWerner replied:
Some of these `City` methods are invoked on the path of loading a scenario, so the singleton isn't necessarily the relevant `GameData` object. Sorry, I noted that in the new commit description but neglected to mention that in the GitHub comment.
